staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go

```diff
@@ -0,0 +1,316 @@
+	hookConfig, err := a.hookSource.Webhooks()
+	// if Webhook configuration is disabled, fail open
+	if err == configuration.ErrDisabled {
+		return &v1alpha1.MutatingWebhookConfiguration{}, nil
+	}
+	if err != nil {
+		e := apierrors.NewServerTimeout(attr.GetResource().GroupResource(), string(attr.GetOperation()), 1)
+		e.ErrStatus.Message = fmt.Sprintf("Unable to refresh the Webhook configuration: %v", err)
+		e.ErrStatus.Reason = "LoadingConfiguration"
+		e.ErrStatus.Details.Causes = append(e.ErrStatus.Details.Causes, metav1.StatusCause{
+			Type:    "MutatingWebhookConfigurationFailure",
+			Message: "An error has occurred while refreshing the MutatingWebhook configuration, no resources can be created/updated/deleted/connected until a refresh succeeds.",
+		})
+		return nil, e
+	}
+	return hookConfig, nil
+}
+
+// Admit makes an admission decision based on the request attributes.
+func (a *MutatingWebhook) Admit(attr admission.Attributes) error {
+	hookConfig, err := a.loadConfiguration(attr)
+	if err != nil {
+		return err
+	}
+	hooks := hookConfig.Webhooks
+	ctx := context.TODO()
+
+	var relevantHooks []*v1alpha1.Webhook
+	for i := range hooks {
+		call, err := a.shouldCallHook(&hooks[i], attr)
+		if err != nil {
+			return err
+		}
+		if call {
+			relevantHooks = append(relevantHooks, &hooks[i])
+		}
+	}
+
+	if len(relevantHooks) == 0 {
+		// no matching hooks
+		return nil
+	}
+
+	// convert the object to the external version before sending it to the webhook
+	versionedAttr := versioned.Attributes{
+		Attributes: attr,
+	}
+	if oldObj := attr.GetOldObject(); oldObj != nil {
+		out, err := a.convertor.ConvertToGVK(oldObj, attr.GetKind())
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+		versionedAttr.OldObject = out
+	}
+	if obj := attr.GetObject(); obj != nil {
+		out, err := a.convertor.ConvertToGVK(obj, attr.GetKind())
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+		versionedAttr.Object = out
+	}
+
+	for _, hook := range relevantHooks {
+		t := time.Now()
+		err := a.callHook(ctx, hook, versionedAttr)
+		admission.Metrics.ObserveWebhook(time.Since(t), err != nil, hook, attr)
+		if err == nil {
+			continue
+		}
+
+		ignoreClientCallFailures := hook.FailurePolicy != nil && *hook.FailurePolicy == v1alpha1.Ignore
+		if callErr, ok := err.(*webhookerrors.ErrCallingWebhook); ok {
+			if ignoreClientCallFailures {
+				glog.Warningf("Failed calling webhook, failing open %v: %v", hook.Name, callErr)
+				utilruntime.HandleError(callErr)
+				continue
+			}
+			glog.Warningf("Failed calling webhook, failing closed %v: %v", hook.Name, err)
+		}
+		return apierrors.NewInternalError(err)
+	}
+
+	// convert attr.Object to the internal version
+	return a.convertor.Convert(versionedAttr.Object, attr.GetObject())
+}
+
+func (a *MutatingWebhook) shouldCallHook(h *v1alpha1.Webhook, attr admission.Attributes) (bool, *apierrors.StatusError) {
+	var matches bool
```

Comment:
// TODO: factor into a common place along with the validating webhook version.

Reply:
+1

Reply:
Done.